extensions/2.0/Vendor/KITTYCAD_boundary_representation/README.md

```diff
@@ -0,0 +1,422 @@
+
+Solids are instantiated in the same way as meshes; however, applications are free to process either the mesh or the B-rep. In this case, `"KITTYCAD_boundary_representation"` must appear under `extensionsUsed` but not under `extensionsRequired`.
+
+#### Instantiation of a solid without a mesh approximation
+
+```json
+{
+    ...,
+    "nodes": [
+        {
+            "extensions": {
+                "KITTYCAD_boundary_representation": {
+                    "solid": 0
+                }
+            }
+        }
+    ]
+    "extensionsUsed": ["KITTYCAD_boundary_representation"],
+    "extensionsRequired": ["KITTYCAD_boundary_representation"],
+    "extensions": {
+        "KITTYCAD_boundary_representation": {
+            "solids": [
+                {
+                    "name": "a B-rep expressing an exact solid volume",
+                    ...
+                }
+            ],
+            ...
+        }
+    }
+}
+```
+
+The file size may be reduced by omitting the mesh approximation. In this case, `"KITTYCAD_boundary_representation"` must appear under both `extensionsUsed` and `extensionsRequired`. Applications process B-rep data directly.
+
+Note: conversion from B-rep to mesh is a non-trivial operation. Mesh approximations should be provided for optimal interoperability.
+
+#### Root extension structure
+
+Boundary representation data extends the root data structure of the JSON portion of the glTF document. Each entry at the root may be referenced by index by other data structures.
+
+```json
+{
+    ...,
+    "extensions": {
+        "KITTYCAD_boundary_representation": {
+            "solids": [],
+            "shells": [],
+            "faces": [],
+            "loops": [],
+            "edges": [],
+            "vertices": []
+        }
+    }
+}
+```
+
+The `shells`, `faces`, `loops`, and `edges` arrays define _orientable_ objects. An important detail of this extension is that orientable objects may be referenced in their reverse-sense using a negative index.
+
+```json
+{
+    ...,
+    "extensions": {
+        "KITTYCAD_boundary_representation": {
+            ...,
+            "faces": [
+                {
+                    "outerLoop": 0,
+                    "surface": 0
+                }
+            ],
+            "loops": [
+                {
+                    "edges": [0, 1, -2]
+                }
+            ],
+            "edges": [
+                {
+                    "start": 0,
+                    "end": 1,
+                    "curve": 0
+                },
+                {
+                    "start": 1,
+                    "end": 2,
+                    "curve": 1
+                },
+                {
+                    "start": 0,
+                    "end": 2,
+                    "curve": -2
+                }
+            ],
+            "curves": [
+                {
+                    "type": "line",
+                    "start": [-1, -1, 0],
+                    "end": [1, -1, 0]
```

Comment:
Possibly. The idea here is that straight edges are a portion of a line curve. For example, an edge may span from vertex [-1, -1, -1] to vertex [1, 1, 1], but be associated with a line curve starting from [-2, -2, -2] in the direction [1, 1, 1].

Reply:
Perhaps it could be either an indexed vertex or an explicit co-ordinate.

```json
{
    "curves": [
        {
            "type": "line",
            "line": {
                "start": 0,
                "end": 1
            }
        },
        {
            "type": "line",
            "line": {
                "start": [0.0, 0.0, 0.0],
                "end": [1.0, 1.0, 1.0]
            }
        }
    ]
}
```

Reply:
Line curves have been simplified to a single form: an origin plus a normalised direction.